### PR TITLE
FI-51 contour plot fill

### PR DESCRIPTION
**Pull request by monfera**
Changes: 

Related to contourgl fill:
- changes to `gl-vis/gl-contour2d`
  - triangle fix from @mikolalysenko
  - use fill color
  - include packages
  - changed example to more fully exercise the code
- changes to `plotly.js`
  - use `gl-contour2d` 1.1.2
  - changes to `convert.js` for the color fill setup
  - jasmine test cases (they only expect fill contours to work without raising an error)

Related to preparation for contourgl fill work:
- adding test utils for setup / teardown (removed for the time being, in favor of a [TBD solution]( https://github.com/plotly/plotly.js/issues/529))
- refactoring of `gl_plot_interact_basic_test` using `Plotly.purge` (residual todo item from the already merged FI-30)

Planned or possible follow-up PRs once the `gl-contour2d` tessellation issue is solved (post the current v1.1.2):
- bump `gl-contour2d` version in `package.json`
- image based test cases for `contourgl` fill / line (now it would 'lock in' problematic artifacts)
- use of more standard surface test patterns (2D wave interference, saddle, noise etc.)
- improving similarity of appearance with the non-GL `contour` e.g. line thickness/color
- consider oversampling for a much finer grid, because `gl-contour2d` doesn't handle curves/splines, and the SVG code isn't directly reusable because the SVG implementation properly relies on splines achievable with the `path` mini-language of the `d` attribute
- depending on the solution to the tessellation issue as well as the possible oversampling above, consider refactoring opportunity for sharing more code between GL `convert` and the non-GL `contour`

### CODE REVIEW

**Comment by monfera** [comment]
image

File: src/traces/contourgl/convert.js [hunk 1]
```diff
@@ -95,9 +96,22 @@ proto.update = function(fullTrace, calcTrace) {
     this.contourOptions.x = this.heatmapOptions.x = calcPt.x;
     this.contourOptions.y = this.heatmapOptions.y = calcPt.y;
 
-    var colorOptions = convertColorscale(fullTrace);
-    this.contourOptions.levels = colorOptions.levels;
-    this.contourOptions.levelColors = colorOptions.levelColors;
+
+    // pass on fill information
+    if(fullTrace.contours.coloring === 'fill') {
+        colorOptions = convertColorscale(fullTrace, true);
```
Comment (etpinard) [inline src/traces/contourgl/convert.js]: I'm _ok_ with the boolean `fill` option here, as `convertColorscale` is only used in this file. But @monfera, I hope you agree that `convertColorscale(fullTrace, { fill: true });` would have been cleaner and more robust.

File: src/traces/contourgl/convert.js [hunk 2]
```diff
@@ -124,20 +138,20 @@ function flattenZ(zIn, rowLen, colLen) {
...
+function convertColorscale(fullTrace, fill) {
     var contours = fullTrace.contours,
         start = contours.start,
         end = contours.end,
         cs = contours.size || 1;
 
     var colorMap = makeColorMap(fullTrace);
 
-    var N = Math.floor((end - start) / cs) + 1,
+    var N = Math.floor((end - start) / cs) + (fill ? 2 : 1), // for K thresholds (contour linees) there are K+1 areas
```
Comment (etpinard) [inline src/traces/contourgl/convert.js]: nicely done!

File: src/traces/contourgl/convert.js (hunk 1 above)
Comment (monfera) [inline src/traces/contourgl/convert.js]: @etpinard Glad to make this small change!

File: test/jasmine/tests/gl2d_scatterplot_contour_test.js [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    var y = d3.range(-1, 1.5, 0.5); // left closed, right open interval
+    var i, j, p, z = new Array(x.length);
+    for(i = 0; i < x.length; i++) {
+        z[i] = new Array(y.length);
+        for(j = 0; j < y.length; j++) {
+            p = rotate(Math.PI / 4, {x: x[i], y: -y[j]});
+            z[i][j] = jitter(maxJitter, Math.pow(p.x, 2) + Math.pow(p.y, 2));
+        }
+    }
+    return {x: x, y: y, z: z}; // looking forward to the ES2015 return {x, y, z}
```
Comment (etpinard) [inline test/jasmine/tests/gl2d_scatterplot_contour_test.js]: :wink:

File: test/jasmine/tests/gl2d_scatterplot_contour_test.js [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        .then(null, failTest)
+        .then(done);
+}
+
+describe('contourgl plots', function() {
+
+    var gd;
+
+    beforeEach(function() {
+        gd =createGraphDiv();
```
Comment (etpinard) [inline test/jasmine/tests/gl2d_scatterplot_contour_test.js]: Hmm. Strange. `npm run lint` should have caught this.

File: test/jasmine/tests/gl2d_scatterplot_contour_test.js [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+            'width': 600,
+            'autosize': true
+        }
+    };
+};
+
+
+function makePlot(gd, mock, done) {
+    return Plotly.plot(gd, mock.data, mock.layout)
+        .then(null, failTest)
```
Comment (etpinard) [inline test/jasmine/tests/gl2d_scatterplot_contour_test.js]: Very well played here.

File: test/jasmine/tests/gl_plot_interact_basic_test.js [hunk 6]
```diff
@@ -1,101 +1,83 @@
...
 
-        function testEvents(plot, done) {
-            plot.then(function(graphDiv) {
-                var tuple = addEventCallback(graphDiv); // TODO disuse tuple with ES6
-                verifyInteractionEffects(tuple);
-                teardown(graphDiv, done);
-            });
-        }
+    afterEach(function() {
+        Plotly.purge(gd);
```
Comment (etpinard) [inline test/jasmine/tests/gl_plot_interact_basic_test.js]: :tada:

File: src/traces/contourgl/convert.js (hunk 1 above)
Comment (etpinard) [inline src/traces/contourgl/convert.js]: No need for now. :+1:

**Comment by etpinard** [comment]
Great. This is a solid patch for `contourgl`. Thank you very much for your efforts @monfera :beers:

**Pull request merged**